[Docs] Advanced networking documentation (#3166)
liqotech/liqo: docs/architecture/network/basic.md

# Basic Concepts

Liqo network is a cloud-native solution that enables seamless communication between Kubernetes clusters. In particular, it allows pod-to-pod communication across clusters. This means that the **services** and **endpointslices** reflection is not managed by the networking module (it is managed by the virtual-kubelet).

The two main areas of the Liqo network are:

- **Internal network**: it manages the traffic inside the cluster and forwards it between nodes and gateways. It uses the Geneve protocol to encapsulate the traffic (the green tunnels in the image below).
- **External network**: it manages the traffic between different clusters. It uses WireGuard to create a secure tunnel between gateways (the orange tunnels in the image below).

![basic](../../_static/images/architecture/network/basev2.excalidraw.svg)

## Internal network

> In the internal network, every node is connected to each gateway (one for each peering) using a dedicated Geneve tunnel. This architecture ensures that:

Reviewer: ```suggestion
In the internal network, each node is connected to the pod acting as **cluster gateway** (one for each peering) using a dedicated Geneve tunnel.
As a consequence, the number of Geneve tunnels in Liqo is equal to the number of _nodes_, and it does not depend on the number of _pods_.
This architecture ensures that:
```

- For each node, there is a dedicated Geneve interface on every gateway.
- For each gateway, there is a dedicated Geneve interface on every node.

> Each Geneve interface is assigned an IP address from the **internal CIDR**. This allows the interfaces to be reachable and enables you to ping the Geneve interfaces both from the node and from the gateway, facilitating connectivity checks and troubleshooting.

Reviewer: ```suggestion
Each Geneve interface is assigned an IP address from the **internal CIDR**.
This allows the interfaces to be reachable and enables you to ping the Geneve interfaces both from the node and from the gateway, facilitating connectivity checks and troubleshooting.
```

This behavior depends on the CNI's handling of inter-node communication and is beyond the scope of Liqo.

#### Known Limitations: Geneve and Kubernetes Services

Technologies that leverage UDP tunnels, such as Geneve, do not support the traversal of intermediate NATs.
The reason can be found on the [RFC 7348](https://www.rfc-editor.org/rfc/rfc7348.html):

> It is recommended that the UDP source port number
> be calculated using a hash of fields from the inner packet --
> one example being a hash of the inner Ethernet frame's headers.
> This is to enable a level of entropy for the ECMP/load-
> balancing of the VM-to-VM traffic across the VXLAN overlay.

In other words, Linux performs _source port hashing_ when encapsulating the pod traffic in the Geneve tunnel, meaning the source port in the UDP header is ephemeral and is calculated per-packet based on 5-tuple of the internal packet.
Therefore, traffic tunneled in UDP cannot traverse the NAT because the UDP 5-tuple in the incoming/outgoing directions are not equal (due to the UDP source port hashing), hence the tuple cannot be found in the connection tracking table of the NAT.

> Since Kubernetes Services are implemented with DNAT rules and Connection Tracking (hence, are not associated to any network interface), Geneve tunnels cannot support any kind of Kubernetes Service as endpoints (ClusterIP, NodePort, or LoadBalancer), but must be directed to a _physical_ IP of the pod/node, which is guaranteed to handled by the CNI without any NAT mechanism.

Reviewer: ```suggestion
Since Kubernetes Services are implemented with DNAT rules and Connection Tracking (hence, are not associated to any network interface), Geneve tunnels cannot support any kind of Kubernetes Service as endpoint (ClusterIP, NodePort, or LoadBalancer), but must be directed to the _real_ IP of the pod/node, which is guaranteed to handled by the CNI without any NAT mechanism.
```

Each WireGuard tunnel is created inside a dedicated gateway pod. For every peering, a dedicated gateway pod is deployed in both clusters involved in the connection. Each gateway pod establishes a secure WireGuard tunnel that connects only to a single remote cluster. This means that every gateway can point to just one remote cluster, ensuring isolation and security for each inter-cluster connection.

There are two types of gateway pods: **gateway-client** and **gateway-server**.

- The **gateway-client** acts as a WireGuard client, initiating the connection to the remote cluster.
- The **gateway-server** acts as a WireGuard server, accepting incoming connections from remote clients. The gateway-server must expose itself using a Kubernetes Service, which can be of type LoadBalancer, NodePort, or ClusterIP, depending on the cluster's networking setup and requirements.

### Gateway Details

Each gateway in Liqo is equipped with at least three distinct interfaces:

- The default interface, which serves as the direct link to external networks.
- A GENEVE interface for each node within the cluster (at least one in the case of a single-node cluster).

> - A tunnel interface for communication with the remote Gateway, WireGuard by default.

Reviewer: ```suggestion
- A tunnel interface to the remote Gateway, using WireGuard by default.
```

### Recap of Packets flow

The flow of a packet from a pod in one cluster to a pod in another cluster involves several steps:

1. The packet originates from a pod and reaches its node through a virtual Ethernet interface. This is managed by the CNI and is standard Kubernetes networking behavior (Liqo is not responsible for this part).
2. From the node, the packet enters a Geneve tunnel using a specific [route](routeconfiguration.md#remote-cluster-id-node-gw-node).The packet arrives at the other end of the Geneve tunnel, inside the gateway pod.
3. The gateway does DNAT and routes the packet into the WireGuard tunnel([check [this route](routeconfiguration.md#remote-cluster-id-gw-ext-gateway)]).
4. The traffic reaches the gateway pod in the remote cluster. It is decrypted by the WireGuard driver and reinserted into the stack. It is then routed to the correct Geneve interface to reach the node hosting the target pod.This is achieved using the following [routes](routeconfiguration.md#local-cluster-id-node-name-gw-node-gateway). After routing, SNAT is applied.
5. The packet is then forwarded from the node to the target pod, leveraging the CNI.

> The return traffic follows the same path in reverse, ensuring symmetric routing between the two pods across clusters.

Reviewer: ```suggestion
The return traffic follows the same path in opposite order, ensuring symmetric routing between the two pods across clusters.
```

![Packets flow](../../_static/images/architecture/network/baseflow-v2.excalidraw.svg)

## CIDR Remapping

> Liqo enables two clusters to peer even if they have the same pod CIDR. Each cluster can independently remap the remote cluster’s pod CIDR to a different, locally chosen CIDR. This remapped CIDR is decided by each cluster on its own and is not shared with other clusters.

Reviewer: ```suggestion
Liqo enables two clusters to peer even if they have the same pod CIDR.
Each cluster can independently remap the remote cluster’s pod CIDR to a different, locally chosen CIDR.
This remapped CIDR is decided locally by each cluster and is not shared, nor need to be negotiated, with other clusters.
```

For example, if both Cluster A and Cluster B use `10.0.0.0/16` as their pod CIDR, Cluster A can remap Cluster B’s pod CIDR (`10.0.0.0/16`) to a new CIDR, such as `11.0.0.0/16`. As a result, Cluster A will be able to reach a pod in Cluster B with IP `10.0.0.6` using the remapped IP `11.0.0.6`. The same applies in the opposite direction if Cluster B chooses to remap Cluster A’s CIDR.

When a packet directed to a remote pod enters the gateway, its destination IP is the remapped one. The gateway performs a DNAT operation to replace this remapped address (used only within the local cluster) with the corresponding real IP address used by the remote cluster.
![Packets flow](../../_static/images/architecture/network/SNAT-DNAT.svg)

> In the example above a packet originating from a local pod destined for a remote pod enters the Gateway with the IP address `11.0.0.6` and will be translated to `10.0.0.6` by the Gateway before being sent to the remote cluster.

Reviewer: ```suggestion
In the above example, a packet originating from a local pod and directed to a remote pod enters the Gateway with the IP address `11.0.0.6`, which will be translated to `10.0.0.6` by the Gateway before sending it to the remote cluster.
```

> In a similar way, when the gateway receives incoming traffic, its source IP is the real one used by the remote cluster, so it is SNATed to the remapped one before being sent to the appropriate pod.

Reviewer: ```suggestion
However, the DNAT on the outgoing traffic is not enough to guarantee a successful communication to the remote pod.
In fact, when the gateway on the other side of the Wireguard tunnel receives the above packet as an incoming traffic, it applies a SNAT rule to the packet, hence remapping the source address (which still belongs to the Pod CIDR of the first cluster) into the proper remapped Pod CIDR of the first cluster.
Then, the packet continues its journey to the final pod.

Liqo clearly splits the responsibilities of the DNAT and SNAT steps (which are both required when two networks with overlapped IP addresses needs to be connected seamlessly without IP renumbering) into the two involved gateway pods: the _exiting_ gateway performs DNAT, the _entering_ gateway performs SNAT.
This splitting of responsibilities allow both gateways to run independently, without having to coordinate themselves about the addresses used in the IP remapping process, hence simplifying the overall management of this process.
```

All the remapping logic is managed by the IPAM component of liqo, which tracks the used IPs and CIDRs inside the cluster.

All the firewall rules for managing this remapping is handled inside the gateway pods, ensuring seamless communication even in the presence of overlapping CIDRs.

> The same logic is applied also to the **external CIDR**. This means that every cluster will be able to remap its neighboring cluster's external CIDR to a different, locally chosen CIDR. Pay attention that if you don't change the default external CIDR used by Liqo (check the helm values), every cluster will use the same external CIDR and neighbour's external CIDRs will appear always remapped.

Reviewer: ```suggestion
The same logic is applied also to the **External CIDR**.
This means that each cluster will be able to remap its neighboring cluster's External CIDR to a different, locally chosen CIDR.
Pay attention that if you do not change the default External CIDR used by Liqo (check the Helm values), each cluster will use the same External CIDR, hence neighbour's external CIDRs will appear always remapped.
```

## Example

> A full example of the entire path can be found [here](./basic_example.md)

Reviewer: ```suggestion
A complete example of the entire journey of a packet can be found [here](./basic_example.md)
```

4. From the node, the packet follows the normal TCP/IP stack processing and exits from the local cluster. This step is managed by the CNI and follows the standard behavior of Kubernetes networking (Liqo is not involved in this step).
5. The tunneled Wireguard traffic reaches the gateway pod in the remote cluster. It is decrypted by the WireGuard driver and re-injected into the stack, within the pod itself. Then, it is routed to the Geneve interface of the tunnel that terminates on the node hosting the target pod. This is achieved using the following [routes](routeconfiguration.md#local-cluster-id-node-name-gw-node-gateway). After routing, and in case the local and remote clusters have overlapped Pod CIDRs, SNAT is applied.
6. Once reached the target node, the packet exits from the Geneve tunnel and is then forwarded to the target pod. This step is managed by the CNI and follows the standard behavior of Kubernetes networking (Liqo is not involved in this step).

The return traffic follows the same path in opposite order, ensuring symmetric routing between the two pods across clusters.

![Packets flow](../../_static/images/architecture/network/baseflow-v2.excalidraw.svg)

> 

Reviewer: ```suggestion

### TCP Clamping
TODO
For more information, check the _TCP Clamping_ section on this [documentation](https://www.cloudflare.com/learning/network-layer/what-is-mss/) provided by Cloudflare.

```

Then, the packet continues its journey to the final pod.

Liqo clearly splits the responsibilities of the DNAT and SNAT steps (which are both required when two networks with overlapped IP addresses needs to be connected seamlessly without IP renumbering) into the two involved gateway pods: the _exiting_ gateway performs DNAT, the _entering_ gateway performs SNAT.
This splitting of responsibilities allow both gateways to run independently, without having to coordinate themselves about the addresses used in the IP remapping process, hence simplifying the overall management of this process.
In fact, all the remapping logic is managed locally by the Liqo IPAM component, which tracks the used IPs and CIDRs inside the cluster, and therefore can simply decide which IP address ranges could be used in the remapping process.

All the firewall rules for managing this remapping is handled inside the gateway pods, ensuring seamless communication even in the presence of overlapping CIDRs.

> The same logic is applied also to the **External CIDR**.

Reviewer: Here we start talking about External CIDR, but:
1) we didn't mention what this is
2) the explanation is very basic, and to me very difficult to understand. I would suggest to expand a little more, and explain better why we do have to change the default setting of the External CIDR in the Helm values.

As a consequence, the number of Geneve tunnels in Liqo is equal to the number of _nodes_, and it does not depend on the number of _pods_.

Each Geneve interface is assigned an IP address from the **internal CIDR**.
This allows the interfaces to be reachable and enables you to ping the Geneve interfaces both from the node and from the gateway, facilitating connectivity checks and troubleshooting.

### Details on GENEVE

[Geneve](https://github.com/torvalds/linux/blob/master/drivers/net/geneve.c) is used to traverse the CNI and reach the Gateway efficiently, without altering the original traffic generated by the pod.
The GENEVE interface encapsulates Ethernet traffic into a UDP datagram without encryption (Ethernet over IP).
![basic](../../_static/images/architecture/network/doc-Geneve.drawio.svg)

The external IP packet has the node's address as its source address and the gateway as its destination address. The destination port is, by default for Liqo, 6091.
Once encapsulated, the packet can be processed by the CNI without issues, as its destination is clearly defined.

> The encapsulated traffic is taken by the CNI and routed to the Gateway, as formally this traffic looks like a packet generated by the note itself.

Reviewer 2: ```suggestion
The encapsulated traffic is taken by the CNI and routed to the Gateway, as formally this traffic looks like a packet generated by the node itself.
```

6. Once reached the target node, the packet exits from the Geneve tunnel and is then forwarded to the target pod. This step is managed by the CNI and follows the standard behavior of Kubernetes networking (Liqo is not involved in this step).

The return traffic follows the same path in opposite order, ensuring symmetric routing between the two pods across clusters.

![Packets flow](../../_static/images/architecture/network/baseflow-v2.excalidraw.svg)

### TCP Clamping

When the IP traffic traverses a tunnel, there is always the risk that large IP packets need to be fragmented before encapsulating them into the tunnel.
Unfortunately, not all the packets can be fragmented (e.g., the ones that have the IP _Don't Fragment_ flag turned on), and in this case the packet is dropped.
	

> Although Liqo does it best to limit this problem, e.g., by reducing the MTU at the traffic source, this is not always effective.

Reviewer 2: We never did it, and I don't even think it's possible
